### PR TITLE
[Assembly] Properties rework for Global placement

### PR DESCRIPTION
**Pull request by PaddleStroke**
image

This is changing the properties of the joint objects from 'Obejct1/2' and 'Part1/2' to a single PropertyXLinkSubHidden 'Reference1/2' that store the full path to the object. With the full path to the object we can now get the real getGlobalPlacement.

This will solve the global placement issues where two identical objects could not be differentiated.

Fix https://github.com/FreeCAD/FreeCAD/issues/15090
Fix https://github.com/FreeCAD/FreeCAD/issues/14706
Fix https://github.com/FreeCAD/FreeCAD/issues/14923

### CODE REVIEW

**Comment by PaddleStroke** [comment]
Currently this is not working with Bodies because of https://github.com/FreeCAD/FreeCAD/issues/15430

File: src/Mod/Assembly/App/AssemblyObject.cpp [hunk 1]
```diff
@@ -1302,29 +1333,32 @@ void AssemblyObject::getRackPinionMarkers(App::DocumentObject* joint,
...
 
-    App::DocumentObject* part1 = getObjFromProp(joint, "Part1");
-    App::DocumentObject* obj1 = getObjFromProp(joint, "Object1");
+    App::DocumentObject* part1 = getMovingPartFromRef(joint, "Reference1");
+    App::DocumentObject* obj1 = getObjFromRef(joint, "Reference1");
     Base::Placement plc1 = getPlacementFromProp(joint, "Placement1");
 
-    App::DocumentObject* part2 = getObjFromProp(joint, "Part2");
-    App::DocumentObject* obj2 = getObjFromProp(joint, "Object2");
+    App::DocumentObject* part2 = getMovingPartFromRef(joint, "Reference2");
```
Comment (chennes) [inline src/Mod/Assembly/App/AssemblyObject.cpp:1340]: This is no longer used -- is that correct?

File: src/Mod/Assembly/App/AssemblyObject.cpp [hunk 2]
```diff
@@ -2185,10 +2187,177 @@ App::DocumentObject* AssemblyObject::getObjFromProp(App::DocumentObject* joint,
...
+    }
+
+    App::Document* doc = prop->getValue()->getDocument();
+
+    std::vector<std::string> names = getSubAsList(prop);
+
+    // Lambda function to check if the typeId is a BodySubObject
+    auto isBodySubObject = [](App::DocumentObject* obj) -> bool {
+        // getViewProviderName instead of isDerivedFrom to avoid dependency on sketcher
+        return (obj->getViewProviderName() == "SketcherGui::ViewProviderSketch"
```
Comment (chennes) [inline src/Mod/Assembly/App/AssemblyObject.cpp:2203]: This is a char* to char* comparison, you can't use operator==.

File: src/Mod/Assembly/App/AssemblyObject.cpp [hunk 3]
```diff
@@ -2185,10 +2187,177 @@ App::DocumentObject* AssemblyObject::getObjFromProp(App::DocumentObject* joint,
...
+                || obj->isDerivedFrom<PartApp::Datum>()  // PartDesign::Point + Line + Plane +
+                                                         // CoordinateSystem
+        );
+    };
+
+    for (size_t i = 0; i < names.size(); ++i) {
```
Comment (chennes) [inline src/Mod/Assembly/App/AssemblyObject.cpp:2209]: This loop is pretty ugly -- lots of array accesses to i+1, etc., which is begging for an off-by-one error or a buffer overflow. Consider refactoring using safer idioms.

File: src/Mod/Assembly/App/AssemblyObject.cpp (hunk 1 above)
Comment (PaddleStroke) [reply src/Mod/Assembly/App/AssemblyObject.cpp:1340]: Indeed you are right!

File: src/Mod/Assembly/App/AssemblyObject.cpp (hunk 2 above)
Comment (PaddleStroke) [reply src/Mod/Assembly/App/AssemblyObject.cpp:2203]: Thanks for the catch !

**Comment by PaddleStroke** [comment]
@chennes If you have some time to check this out that would be awesome :)

**Comment by PaddleStroke** [comment]
I just dropped the commit

> Assembly: AssemblyObject : add PropertyBool 'Static'

Since it was out of subject of this PR.

File: src/Mod/Assembly/App/AssemblyObject.h [hunk 4]
```diff
@@ -57,6 +62,8 @@ namespace Assembly
 class JointGroup;
 class ViewGroup;
 
+using ObjRefPair = std::pair<App::DocumentObject*, App::PropertyXLinkSub*>;
```
Comment (chennes) [inline src/Mod/Assembly/App/AssemblyObject.h:65]: I don't love the use of std::pair here (it makes later code quite opaque, as the reader has to remember what `.first` and `.second` refer to. This seems to get used often enough to justify being a struct, in my opinion.

**Review by chennes — COMMENTED** [review]
This compiles without warnings on MSVC, and does not introduce any additional linter issues.

File: src/Mod/Assembly/App/AssemblyObject.h (hunk 4 above)
Comment (PaddleStroke) [reply src/Mod/Assembly/App/AssemblyObject.h:65]: Sounds reasonable.

Comment (PaddleStroke) [reply src/Mod/Assembly/App/AssemblyObject.h]: Done

**Comment by s-quirin** [comment]
@PaddleStroke I used the UtilsAssembly.getGlobalPlacement(targetObj) before this PR. How can I get reference from my object to use the new UtilsAssembly.getGlobalPlacement(ref)?

**Comment by PaddleStroke** [comment]
@s-quirin I need more context.
The ref is a PropertyXLinkSub with the complete subnames. There should be a migration script for normal assembly usage. But I guess you are using the getGlobalPlacement is a macro or something?

**Comment by s-quirin** [comment]
@PaddleStroke Yes, I use it in a macro to get the global placement of Links. I found the answer in `JointObject.handleInitialSelection()` and have figured out that I could use it like this:
```
import UtilsAssembly
selection = Gui.Selection.getSelection()
sel = selection[0]
# placement = UtilsAssembly.getGlobalPlacement(sel)    # old version
ref = [sel.getParent(), [sel.Name + '.']]
if ref[0] is None:
    ref[0] = sel

placement = UtilsAssembly.getGlobalPlacement(ref)
```
But it doesn't work for Links on the root level: Placement is shifted by a factor of 2.
For now, I'm sticking with the old version of your function which can handle selections on the root level correctly.

**Comment by PaddleStroke** [comment]
FIY there's a PR modifying getGlobalPlacement already : https://github.com/FreeCAD/FreeCAD/pull/15262#issuecomment-2308397269
I'm not sure it's solving the root level question though.

**Comment by Rexbas** [comment]
> But it doesn't work for Links on the root level: Placement is shifted by a factor of 2.

I think I had the same or a similar problem which I described here: https://github.com/FreeCAD/FreeCAD/pull/15262#issuecomment-2267435630
This should be fixed in the cpp version in that PR. 

@PaddleStroke, isn´t it possible to let the python version call the cpp version? It would be easier to maintain only one version. I believe there are some more issues with links, regarding the link transform = true property, but I will investigate that later and make a separate issue for that when I am sure.

**Comment by PaddleStroke** [comment]
It would be much better not to have a python version and a cpp version.
If you can remove the python version it would be much appreciated.

The problem I had was to make the python binding of the cpp version. But I have not looked much into it.

**Comment by Rexbas** [comment]
I don't know how python bindings work either but maybe I will look into it.

**Pull request merged**